src/main/kotlin/org/rust/ide/annotator/fixes/AddTurbofishFix.kt

```diff
@@ -3,55 +3,103 @@ package org.rust.ide.annotator.fixes
-        } else {
-            base.parentOfType<RsBinaryExpr>()
+        if (base.left !is RsBinaryExpr) {
+            return resolveMatchExpression(base) ?: base
+        }
+        val left = base.left as RsBinaryExpr
+        if (left.operatorType == GTGT) {
+            return resolveMatchExpression(base) ?: base
         }
+        return base
     }
 
     override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
-        val matchExpr = resolveMatchExpression(element) ?: return null
-        val left = matchExpr.left as? RsBinaryExpr ?: return null
-        if (left.operatorType != LT || matchExpr.operatorType != GT) {
-            return null
+        val m = resolveMatchExpression(element) ?: return null
+        return guessContext(project, m)
+    }
+
+    private fun innerOffset(root: PsiElement, child: PsiElement): Int? {
+        if (child == root) {
+            return 0
+        }
+        val parent = child.parent ?: return null
+        val upper = innerOffset(root, parent) ?: return null
+        return upper + child.startOffsetInParent
+    }
+
+    private fun guessContext(project: Project, binary: RsBinaryExpr): Context? {
+        val nodes = bfsLeafs(binary)
+        val called = rightBoundary(nodes) ?: return null
+        val typeListEndIndex = binary.textLength - called.textLength
+
+        val turbofish_offset = nodes.takeWhile { it != called }.map {
+            val offset = innerOffset(binary, it)!! + it.textLength
+            val typeListCandidate = binary.text.substring(offset, typeListEndIndex)
+            if (isTypeArgumentList(project, typeListCandidate)) {
+                offset
+            } else {
+                null
+            }
+        }.filterNotNull().firstOrNull() ?: return null
+        return Context(binary, turbofish_offset)
+    }
+
+    private fun rightBoundary(nodes: List<RsExpr>) = nodes.asReversed().firstOrNull { isCallExpression(it) }
+
+    private fun isTypeArgumentList(project: Project, candidate: String) =
+        create<RsCallExpr>(project, "something$TURBOFISH$candidate()") != null
+
+    private fun bfsLeafs(expr: RsExpr?): List<RsExpr> {
+        return when (expr) {
+            is RsBinaryExpr -> {
+                bfsLeafs(expr.left) + bfsLeafs(expr.right)
+            }
+            null -> listOf<RsExpr>()
+            else -> listOf(expr)
         }
-        val caller = left.left
-        val gen = left.right as? RsPathExpr ?: return null
-        val more = matchExpr.right ?: return null
-        if (!isCallExpression(more)) {
-            return null
-        };
-        return Context(matchExpr, caller, gen, more)
     }
 
     private fun isCallExpression(expr: RsExpr) = expr is RsParenExpr || expr.firstChild is RsParenExpr
+
+    /* Maybe there' s a better way than cut and paste these methods from RsFactory */
+    private inline fun <reified T : RsCompositeElement> create(project: Project, text: String): T? =
```

Comment:
No, currently `RsFactory` creates only valid elements. You can add a `createDummyFile(text: String)` method to it, but this particular fallible create belongs to this fix.